// y-drawer/src/webcomponent/index.js
// Copyright (c) 2017 Florian Klampfer
// Licensed under MIT

import HTMLYDrawerElement from './y-drawer';

const templateV1 = `
<div class="y-drawer-scrim"></div>
<div class="y-drawer-content">
  <slot></slot>
</div>

<style>
  @media screen {
    .y-drawer-scrim {
      display: block;
      position: fixed;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      opacity: 0;
      z-index: 0;
      pointer-events: none;
      transform: translate3d(0,0,0);
      -webkit-tap-highlight-color: transparent;
  
      background: rgba(0, 0, 0, 0.67);
    }
  
    .y-drawer-content {
      position: fixed;
      left: 0;
      bottom: 0;
      top: 0;
      z-index: 1;
      overflow-x: hidden;
      overflow-y: auto;
      transform: translate3d(0,0,0);
      -webkit-overflow-scrolling: touch;
      contain: strict;
  
      left: -18rem;
      width: 18rem;
      background: #e8e8e8;
    }
  
    .y-drawer-content.y-drawer-opened {
      left: 0!important;
      transform: translate3d(0,0,0)!important;
    }
  }
  
  @media print {
    .y-drawer-scrim {
      display: none!important;
    }
  
    .y-drawer-content {
      transform: none!important;
    }
  }
  
  @media screen {
    :host {
  
    }
  
    .y-drawer-content {
      @apply --y-drawer-content-container;
      width: var(--y-drawer-width, 18rem);
      left: calc(-1 * var(--y-drawer-slide-width, var(--y-drawer-width, 18rem)));
      background: var(--y-drawer-background, #e8e8e8);
    }
  
    .y-drawer-scrim {
      background: var(--y-drawer-scrim-background, rgba(0, 0, 0, 0.67));
    }
  }
  
</style>

`;

const templateV0 = `
<div class="y-drawer-scrim"></div>
<div class="y-drawer-content">
  <content select="*"></content>
</div>

<style>
  @media screen {
    .y-drawer-scrim {
      display: block;
      position: fixed;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      opacity: 0;
      z-index: 0;
      pointer-events: none;
      transform: translate3d(0,0,0);
      -webkit-tap-highlight-color: transparent;
  
      background: rgba(0, 0, 0, 0.67);
    }
  
    .y-drawer-content {
      position: fixed;
      left: 0;
      bottom: 0;
      top: 0;
      z-index: 1;
      overflow-x: hidden;
      overflow-y: auto;
      transform: translate3d(0,0,0);
      -webkit-overflow-scrolling: touch;
      contain: strict;
  
      left: -18rem;
      width: 18rem;
      background: #e8e8e8;
    }
  
    .y-drawer-content.y-drawer-opened {
      left: 0!important;
      transform: translate3d(0,0,0)!important;
    }
  }
  
  @media print {
    .y-drawer-scrim {
      display: none!important;
    }
  
    .y-drawer-content {
      transform: none!important;
    }
  }
  
  @media screen {
    :host {
  
    }
  
    .y-drawer-content {
      @apply --y-drawer-content-container;
      width: var(--y-drawer-width, 18rem);
      left: calc(-1 * var(--y-drawer-slide-width, var(--y-drawer-width, 18rem)));
      background: var(--y-drawer-background, #e8e8e8);
    }
  
    .y-drawer-scrim {
      background: var(--y-drawer-scrim-background, rgba(0, 0, 0, 0.67));
    }
  }
  
</style>

`;

function fragmentFromString(strHTML) {
  return document.createRange().createContextualFragment(strHTML);
}

export default class extends HTMLYDrawerElement {
  getTemplateInstance(version) {
    switch (version) {
      case 'v1': return fragmentFromString(templateV1);
      case 'v0': return fragmentFromString(templateV0);
      default: throw Error();
    }
  }
}
