// y-drawer/src/vanilla/index.js
// Copyright (c) 2017 Florian Klampfer
// Licensed under MIT

/* eslint-disable import/no-extraneous-dependencies, import/no-unresolved, import/extensions */
import drawerCore from '../core';

const style = `
<style>
@media screen {
  .y-drawer-scrim {
    display: block;
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    opacity: 0;
    z-index: 0;
    pointer-events: none;
    transform: translate3d(0,0,0);
    -webkit-tap-highlight-color: transparent;

    background: rgba(0, 0, 0, 0.67);
  }

  .y-drawer-content {
    position: fixed;
    left: 0;
    bottom: 0;
    top: 0;
    z-index: 1;
    overflow-x: hidden;
    overflow-y: auto;
    transform: translate3d(0,0,0);
    -webkit-overflow-scrolling: touch;
    contain: strict;

    left: -18rem;
    width: 18rem;
    background: #e8e8e8;
  }

  .y-drawer-content.y-drawer-opened {
    left: 0!important;
    transform: translate3d(0,0,0)!important;
  }
}

@media print {
  .y-drawer-scrim {
    display: none!important;
  }

  .y-drawer-content {
    transform: none!important;
  }
}

</style>`;


function fragmentFromString(strHTML) {
  return document.createRange().createContextualFragment(strHTML);
}

export default class Drawer extends drawerCore() {
  constructor(el, props) {
    super();
    this.setupComponent(el, props);
  }

  // @override
  setupDOM(el) {
    if (!el) throw Error('No element provided');

    const scrim = document.createElement('div');
    scrim.classList.add('y-drawer-scrim');

    const content = document.createElement('div');
    content.classList.add('y-drawer-content');
    while (el.children.length > 0) {
      content.appendChild(el.children[0]);
    }

    el.appendChild(scrim);
    el.appendChild(content);

    const ref = document.getElementsByTagName('style')[0];
    ref.parentNode.insertBefore(fragmentFromString(style), ref);

    return el;
  }
}
